tests/perf/common/test_harness.h

```diff
@@ -0,0 +1,193 @@
+  using TestResults = std::vector<std::vector<TestResult>>;
+  using PerNodeResults = std::unordered_map<NodeType, TestResultHolder<TimeType>>;
+  using CombinedResults = std::vector<std::pair<TestName, PerNodeResults>>;
+
+  // Memory use at the end of test iteration (i.e. phase)
+  using MemoryUsage = std::vector<std::vector<std::size_t>>;
+  using CombinedMemoryUse = std::unordered_map<NodeType, std::vector<TestResultHolder<std::size_t>>>;
+
+  virtual ~PerfTestHarness() = default;
+
+  virtual void SetUp();
+  virtual void TearDown();
+
+  /**
+   * \brief Initialize internal variables and parse args
+   * Perf specific args:
+   * --vt_perf_gen_file - generate .CSV files with test results
+   * --vt_perf_verbose  - output per iteration times/memory use
+   * --vt_perf_num_runs - set how many times the tests should run
+   *                      (e.g. --vt_perf_num_run=100)
+   */
+  void Initialize(int argc, char** argv);
+
+  /**
+   * \brief Get the name of this test suite
+   *
+   * \return name
+   */
+  std::string GetName() const;
+
+  /**
+   * \brief Get the number of runs that this test will run
+   *
+   * \return Number of runs
+   */
+  uint32_t GetNumRuns() const;
+
+  /**
+   * \brief Dump the test results to stdout and CSV files.
+   * {name_}_mem.csv file which contains memory usage for each iteration
+   * {name_}_time.csv file which contains all time-related data
+   *
+   * This is called at the end of running test suite.
+   */
+  void DumpResults() const;
+
+  /**
+   * \brief Add a single test result (name-time pair)
+   *
+   * \param[in] test_result name-time pair of test result
+   */
+  void AddResult(TestResult const& test_result);
+
+  /**
+   * \brief Add and start a timer with name \c name
+   *
+   * \param[in] name name of the timer
+   */
+  void StartTimer(std::string const& name);
+
+  /**
+   * \brief Stop the timer \c name and add test result.
+   * This function calls \c AddResult
+   *
+   * \param[in] name name of the timer
+   */
+  void StopTimer(std::string const& name);
+
+  /**
+   * \brief Send the tests' results to root node.
+   * This is called after each test run.
+   */
+  void SyncResults();
+
+  /**
+   * \brief Handler for receiving test results from other nodes
+   *
+   * \param[in] msg message that contains the results
+   */
+  static void RecvTestResult(TestMsg* msg);
+
+  /**
+   * \brief Spin the scheduler until all work is done.
+   */
+  static void SpinScheduler();
+
+  /**
+   * \brief Helper function used for tracking memory usage.
+   * Should be called each iteration. Uses \c StatM to track memory usage
+   */
+  void GetMemoryUsage();
+
+  private:
+  std::string OutputMemoryUse() const;
+  std::string OutputTimeResults() const;
+
+  protected:
+  bool gen_file_ = false;
+  bool verbose_ = false;
+  uint32_t num_runs_ = 50;
+  uint32_t current_run_ = 0;
+  std::vector<char*> custom_args_ = {};
+
+  NodeType my_node_ = {};
```

Comment:
Initialize these with the type sentinels:
```suggestion
  NodeType my_node_ = uninitialized_destination;
```